src/sas/sascalc/dataloader/readers/red2d_reader.py

```diff
@@ -236,30 +236,19 @@ def get_file_contents(self):
         ymin = np.min(qy_data)
         ymax = np.max(qy_data)
 
-        ## calculate the range of the qx and qy_data
-        x_size = math.fabs(xmax - xmin)
-        y_size = math.fabs(ymax - ymin)
-
-        # calculate the number of pixels in the each axes
-        npix_y = math.floor(math.sqrt(len(data)))
-        npix_x = math.floor(len(data) / npix_y)
-
-        # calculate the size of bins
-        xstep = x_size / (npix_x - 1)
-        ystep = y_size / (npix_y - 1)
-
         # store x and y axis bin centers in q space
-        x_bins = np.arange(xmin, xmax + xstep, xstep)
-        y_bins = np.arange(ymin, ymax + ystep, ystep)
-
-        # get the limits of q values
-        xmin = xmin - xstep / 2
-        xmax = xmax + xstep / 2
-        ymin = ymin - ystep / 2
-        ymax = ymax + ystep / 2
+        x_bins = np.unique(qx_data)
+        y_bins = np.unique(qy_data)
+        # For non-uniform Qx and Qy arrays
+        if round(len(x_bins) * len(y_bins) / len(qx_data)) >= 2:
+            x_bins = [qx_data[i] for i in takewhile(
+                lambda i: qx_data[i] < qx_data[i + 1], range(len(qx_data)))]
+            # Above list comprehension rejects last item
+            x_bins.append((qx_data[len(x_bins)]))
+            qy = np.reshape(qy_data, (len(qx_data)//len(x_bins), len(x_bins)))
+            y_bins = np.transpose(qy)[0].tolist()
```

Comment:
This code would benefit from some comments describing the data structure that it is operating over and what it is doing.

Nearest I can tell it is assuming that qx and qy are rectilinear row-major arrays, with qx increasing along the rows and qy increasing along the columns. It is using the first row of qx and first column of qy as the bin centers for the entire array.  So you ought to be able to use:
```python
x_bins = qx_data.reshape(num_qy, num_qx)[0, :]
y_bins = qy_data.reshape(num_qy, num_qx)[:, 0]
```
The problems is that you do not know how many qx and qy.

Using `np.unique` is a little risky since the data writer may have introduced spurious differences in the stored values for qx and qy, even though they are nominally the same.  It may be that you end up with a situation where the size of `x_bins * y_bins` does not match the array.  I'm guessing this won't cause runtime errors since sasview is only using x_bins and y_bins as suggested rectilinear bin centers, and will rebin the data onto those centers later, but it does mean that you may end up with some extremely narrow rows/columns.

It took some time to understand what takewhile is doing.

How about you instead set the number of x_bins as the first k such that `qx_data[k] < qx_data[k-1]`, then take the first row:
```
# Assume qx_data and qy_data are flattened rectilinear arrays in row major order with qx_data increasing
# along each row. Set x_bins to the first row of qx and y_bins to the first column of qy_data. You can
# find the end of each row by comparing qx_data[k] < qx_data[k-1].  With argmax you can find
# the index of the first true value, and add one to convert to row length.  Set the last element to True in
# case there is only one row.
num_qx = np.argmax(np.hstack((qx_data[1:] < qx_data[:-1], True)) + 1
num_qy = len(qy_data) // num_qx
assert num_qx * num_qy == len(qy_data), "q data is not rectilinear"
x_bins = qx_data.reshape(num_qy, num_qx)[0, :]
y_bins = qy_data.reshape(num_qy, num_qx)[:, 0]
```

This code will fail if you have a single column qy != 0 with qx values computed for a flat detector, or due to numerical jitter on a spherical detector leading to slightly different qx values in each bin.

If add the additional assumption that `num_qx > 1` and `num_qy > 1` you can remove the `hstack`.

Reply:
Just to be clear, I was proposing that you replace the entire block, from line 240-252, including the np.unique and the conditional test. This will be more robust than using np.unique. It will also be faster since it won't need to do a sort, but that is less important.